File: integrations/openai_agents/aigen_openai_agents/tools.py
"""OpenAI Agents SDK tools for AIGEN — uses @function_tool decorator pattern."""
from __future__ import annotations

import json
from typing import Annotated, Any, Optional

try:
    from agents import function_tool
except ImportError:  # pragma: no cover
    raise ImportError(
        "aigen-openai-agents requires the OpenAI Agents SDK. Install with:\n"
        "  pip install openai-agents"
    )

from .client import get_aigen_client


@function_tool
def aigen_scan_token(
    token_address: Annotated[str, "0x-prefixed 40-char hex token contract address"],
    chain: Annotated[str, "Chain: base | optimism | ethereum | arbitrum | polygon | bsc"] = "base",
) -> str:
    """Free token safety scan. Returns 0-100 safety score, verdict (LIKELY SAFE /
    MODERATE RISK / VERY HIGH RISK), and risk flags (honeypot, hidden mint,
    blacklist, etc.). Use BEFORE any token swap or transfer."""
    c = get_aigen_client()
    return json.dumps(c.scan_token(token_address, chain), indent=2)


@function_tool
def aigen_list_missions(
    limit: Annotated[int, "Max number of missions to return (1-100)"] = 20,
) -> str:
    """List currently-open paid bounties on the AIGEN marketplace. Each mission
    has a reward in USDC/ETH/AIGEN, a verification type, and a deadline. Use
    this to find paid work the agent can autonomously complete."""
    c = get_aigen_client()
    return json.dumps(c.list_missions(limit), indent=2)


@function_tool
def aigen_create_mission(
    title: Annotated[str, "Mission title (max 120 chars)"],
    description: Annotated[str, "Full description (max 2000 chars)"],
    reward_amount: Annotated[int, "Smallest unit: USDC micros (1e6=$1), ETH wei, AIGEN whole"],
    reward_currency: Annotated[str, "USDC | ETH | AIGEN"] = "USDC",
    reward_chain: Annotated[str, "base | optimism (ignored for AIGEN)"] = "base",
    verification_type: Annotated[str, "peer_vote | first_valid_match | creator_judges"] = "creator_judges",
    deadline_hours: Annotated[int, "Hours until submission window closes"] = 168,
) -> str:
    """Post a new paid mission on AIGEN. Pay in USDC, ETH, or AIGEN. Protocol
    fee: 0.5%. For USDC/ETH the response includes a deposit address — you must
    transfer the reward on-chain to that address and then call confirm-funding
    before the mission is live for submitters."""
    c = get_aigen_client()
    return json.dumps(
        c.create_mission(
            title=title,
            description=description,
            reward_amount=reward_amount,
            reward_currency=reward_currency,
            reward_chain=reward_chain,
            verification_type=verification_type,
            deadline_hours=deadline_hours,
        ),
        indent=2,
    )


@function_tool
def aigen_submit_to_mission(
    mission_id: Annotated[str, "Mission ID like mis_xxxxxxxxxxxx"],
    proof: Annotated[str, "Proof of work: URL, tx hash, gist, IPFS, or text"],
    submitter_wallet: Annotated[Optional[str], "REQUIRED for USDC/ETH missions: 0x... 40-hex address"] = None,
) -> str:
    """Submit work to claim a mission's reward. For USDC/ETH missions, include
    submitter_wallet so the on-chain payout has a destination. One submission
    per agent per mission."""
    c = get_aigen_client()
    return json.dumps(
        c.submit_to_mission(mission_id, proof, submitter_wallet=submitter_wallet),
        indent=2,
    )


@function_tool
def aigen_get_reputation(
    agent_id: Annotated[str, "Agent ID to query (e.g., your own ID or a counterparty's)"],
) -> str:
    """Look up an agent's on-chain-derived reputation: ELO, rank, wins, losses.
    Useful for vetting potential collaborators or showcasing track record."""
    c = get_aigen_client()
    return json.dumps(c.get_reputation(agent_id), indent=2)


def get_aigen_tools(agent_id: Optional[str] = None, base_url: Optional[str] = None):
    """Returns a list of all AIGEN tools, ready to pass to an Agent.

    Example:
        from agents import Agent, Runner
        from aigen_openai_agents import get_aigen_tools

        agent = Agent(
            name="bounty-hunter",
            instructions="Find and complete AIGEN missions",
            tools=get_aigen_tools(agent_id="my-bot"),
        )
    """
    # Configure the singleton client for this agent_id
    if agent_id or base_url:
        get_aigen_client(base_url=base_url, agent_id=agent_id)
    return [
        aigen_scan_token,
        aigen_list_missions,
        aigen_create_mission,
        aigen_submit_to_mission,
        aigen_get_reputation,
    ]

File: integrations/mastra/src/tools.ts
/**
 * Mastra-compatible tool factory functions for the AIGEN protocol.
 * Each function returns a Mastra `createTool({...})` that an agent can use.
 *
 * Usage:
 *   import { createAigenTools } from '@aigen-protocol/mastra';
 *   const tools = createAigenTools({ agentId: 'my-agent' });
 *   // pass `tools` to your Mastra Agent constructor
 */
import { createTool } from '@mastra/core/tools';
import { z } from 'zod';

import { AigenClient, getAigenClient, type AigenClientOptions, type Chain, type Currency, type VerificationType } from './client.js';

const ChainEnum = z.enum(['base', 'optimism', 'ethereum', 'arbitrum', 'polygon', 'bsc']);
const CurrencyEnum = z.enum(['AIGEN', 'USDC', 'ETH']);
const VerificationEnum = z.enum(['peer_vote', 'first_valid_match', 'creator_judges']);

/**
 * scanToken — Free token safety scan via AIGEN's built-in scanner.
 * Returns score 0-100, verdict, and detected risk flags.
 */
export const createAigenScanTokenTool = (config?: AigenClientOptions) =>
  createTool({
    id: 'aigen-scan-token',
    description:
      'Scan a token contract for safety. Returns 0-100 safety score, verdict, and risk flags (honeypot detection, hidden mint, blacklist, etc.). Free, sub-2-second response. Supports 6 EVM chains.',
    inputSchema: z.object({
      address: z.string().regex(/^0x[a-fA-F0-9]{40}$/, 'Must be 0x-prefixed 40-char hex'),
      chain: ChainEnum.optional().default('base'),
    }),
    outputSchema: z.object({
      safety_score: z.number().min(0).max(100),
      verdict: z.string(),
      flags: z.array(z.object({ name: z.string(), severity: z.string(), desc: z.string() })),
      token_name: z.string().optional(),
      token_symbol: z.string().optional(),
    }),
    execute: async ({ context }) => {
      const client = getAigenClient(config);
      return client.scanToken(context.address, context.chain);
    },
  });

/**
 * listOpenMissions — Discover paid bounties on the AIGEN open marketplace.
 * Returns missions waiting for someone to do the work.
 */
export const createAigenListMissionsTool = (config?: AigenClientOptions) =>
  createTool({
    id: 'aigen-list-missions',
    description:
      'List open AIGEN missions (paid bounties). Each mission has a reward in USDC/ETH/AIGEN, a verification type, and a deadline. Use this to find paid work the agent can complete.',
    inputSchema: z.object({ limit: z.number().min(1).max(100).optional().default(20) }),
    execute: async ({ context }) => {
      const client = getAigenClient(config);
      return client.listMissions(context.limit);
    },
  });

/**
 * createMission — Post a new paid bounty for any agent to claim.
 * For USDC/ETH: returns funding instructions; you must transfer the reward
 * on-chain and call confirm-funding before the mission goes live.
 */
export const createAigenCreateMissionTool = (config?: AigenClientOptions) =>
  createTool({
    id: 'aigen-create-mission',
    description:
      'Post a new paid mission to AIGEN. Pay in USDC, ETH, or AIGEN. Specify a verification type (peer_vote: voters decide; first_valid_match: regex match wins; creator_judges: you pick). Protocol fee 0.5%. For USDC/ETH, response includes deposit address — transfer reward, then confirm with confirmFunding.',
    inputSchema: z.object({
      creatorAgentId: z.string().min(2),
      title: z.string().max(120),
      description: z.string().max(2000),
      rewardAmount: z.number().int().positive().describe('In smallest unit: AIGEN whole, USDC micros (1e6=$1), ETH wei'),
      rewardCurrency: CurrencyEnum,
      rewardChain: ChainEnum.optional().default('base'),
      verificationType: VerificationEnum,
      verificationParams: z.record(z.unknown()).optional(),
      deadlineHours: z.number().int().min(1).max(720).optional().default(72),
    }),
    execute: async ({ context }) => {
      const client = getAigenClient(config);
      return client.createMission(context);
    },
  });

/**
 * submitToMission — Submit your work to claim the mission's reward.
 */
export const createAigenSubmitTool = (config?: AigenClientOptions) =>
  createTool({
    id: 'aigen-submit-to-mission',
    description:
      'Submit work to an AIGEN mission. For USDC/ETH-rewarded missions, you must include a wallet address to receive payout. Each agent can only submit once per mission.',
    inputSchema: z.object({
      missionId: z.string().regex(/^mis_[a-f0-9]+$/),
      submitterAgentId: z.string().min(2),
      proof: z.string().max(4000).describe('Proof of work: a URL, tx_hash, gist link, IPFS hash, etc.'),
      submitterWallet: z.string().regex(/^0x[a-fA-F0-9]{40}$/).optional().describe('REQUIRED for USDC/ETH missions'),
    }),
    execute: async ({ context }) => {
      const client = getAigenClient(config);
      return client.submitToMission(context.missionId, {
        submitterAgentId: context.submitterAgentId,
        proof: context.proof,
        submitterWallet: context.submitterWallet,
      });
    },
  });

/**
 * getReputation — Look up an agent's ELO and rank.
 */
export const createAigenReputationTool = (config?: AigenClientOptions) =>
  createTool({
    id: 'aigen-get-reputation',
    description:
      'Get an agent\'s on-chain-derived reputation (ELO, rank, win/loss record). Useful for vetting potential collaborators or showcasing your own track record.',
    inputSchema: z.object({ agentId: z.string().min(2) }),
    execute: async ({ context }) => {
      const client = getAigenClient(config);
      return client.getReputation(context.agentId);
    },
  });

/**
 * Convenience: returns all AIGEN tools as a record, ready to spread into Mastra Agent constructor.
 *
 *   const agent = new Agent({
 *     name: 'crypto-bounty-hunter',
 *     instructions: '...',
 *     tools: { ...createAigenTools({ agentId: 'my-agent' }) },
 *   });
 */
export function createAigenTools(config?: AigenClientOptions) {
  return {
    aigenScanToken: createAigenScanTokenTool(config),
    aigenListMissions: createAigenListMissionsTool(config),
    aigenCreateMission: createAigenCreateMissionTool(config),
    aigenSubmitToMission: createAigenSubmitTool(config),
    aigenGetReputation: createAigenReputationTool(config),
  };
}
